### PR TITLE
Implement add Users and Usergroups in Users Tab

### PR DESCRIPTION
**Pull request by subinasr**

### CODE REVIEW

File: src/views/ProjectEdit/Users/UserGroupList/index.tsx [hunk 1]
```diff
@@ -103,6 +112,10 @@ function UserGroupList(props: Props) {
         />
     ), []);
 
+    const handleAddUsergroupClick = useCallback(() => {
+        setModalShow();
+    }, []);
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/index.tsx]: add setModalShow in dependency array

File: src/views/ProjectEdit/index.tsx [hunk 2]
```diff
@@ -42,7 +42,7 @@ function ProjectEdit(props: ViewProps) {
         activeProject,
     } = props;
 
-    const [activeTab, setActiveTab] = useState<TabNames>('general');
+    const [activeTab, setActiveTab] = useState<TabNames>('users');
```
Comment (tnagorra) [inline src/views/ProjectEdit/index.tsx]: Shouldn't the active tab be 'general'?

File: src/views/ProjectEdit/Users/UserList/index.tsx [hunk 3]
```diff
@@ -113,6 +174,10 @@ function UserList(props: Props) {
         />
     ), []);
 
+    const handleAddUserClick = useCallback(() => {
+        setModalShow();
+    }, [setModalShow]);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/index.tsx]:
```suggestion
    const handleAddUserClick = setModalShow;
```

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    MultiResponse,
+} from '#typings';
+import { ProjectRole } from '#typings/project';
+import _ts from '#ts';
+import styles from './styles.scss';
+
+interface Props {
+    onModalClose: () => void;
+    projectId: string;
+    reloadTable: () => void;
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]:
```suggestion
    onTableReload: () => void;
```

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    } = props;
+
+    const defaultFormValues: PartialForm<FormType> = useMemo(() => {
+        if (isDefined(usergroupValue)) {
+            return usergroupValue;
+        }
+        return {};
+    }, [usergroupValue]);
+
+    console.warn(usergroupValue);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: No console.warn meant for debugging

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        reloadTable,
+        usergroupValue,
+    } = props;
+
+    const defaultFormValues: PartialForm<FormType> = useMemo(() => {
+        if (isDefined(usergroupValue)) {
+            return usergroupValue;
+        }
+        return {};
+    }, [usergroupValue]);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: We don't  need this. We can simply do

usergroupValue ?? {};

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            onModalClose();
+        },
+        onFailure: (_, errorBody) => {
+            notifyOnFailure(_ts('projectEdit', 'projectUsergroupPostFailedLabel'))({ error: errorBody });
+        },
+    });
+
+    const handleSubmit = useCallback(() => {
+        triggerAddUserGroup();
+    }, [triggerAddUserGroup]);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]:
```suggestion
    const handleSubmit = triggerAddUserGroup;
```

File: src/views/ProjectEdit/Users/UserGroupList/index.tsx [hunk 8]
```diff
@@ -38,21 +49,58 @@ function UserGroupList(props: Props) {
...
 
     const [activePage, setActivePage] = useState<number>(1);
+    const [usergroupIdToDelete, setUsergroupIdToDelete] = useState<number | undefined>(undefined);
+    const [usergroupIdToEdit, setUsergroupIdToEdit] = useState<number | undefined>(undefined);
     const queryForRequest = useMemo(() => ({
         offset: (activePage - 1) * maxItemsPerPage,
         limit: maxItemsPerPage,
     }), [activePage]);
 
+    console.warn(usergroupIdToEdit);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/index.tsx]: No console.warn meant for debugging

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    MultiResponse,
+} from '#typings';
+import { ProjectRole } from '#typings/project';
+import _ts from '#ts';
+import styles from './styles.scss';
+
+interface Props {
+    onModalClose: () => void;
+    projectId: number;
+    reloadTable: () => void;
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]:
```suggestion
    onTableReload: () => void;
```

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+            onModalClose();
+        },
+        onFailure: (_, errorBody) => {
+            notifyOnFailure(_ts('projectEdit', 'projectMembershipPostFailed'))({ error: errorBody });
+        },
+    });
+
+    const handleSubmit = useCallback(() => {
+        triggerAddProjectMember();
+    }, [triggerAddProjectMember]);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]:
```suggestion
    const handleSubmit = triggerAddProjectMember();
```

**Review by samshara — COMMENTED** [review]
please fix deepscan issues

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
+import React from 'react';
+
+import {
+    isDefined,
+    // isNotDefined,
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: remove this if not used

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        onFailure: (_, errorBody) => {
+            notifyOnFailure(_ts('projectEdit', 'projectUsergroupPostFailedLabel'))({ error: errorBody });
+        },
+    });
+
+    const handleSubmit = triggerAddUserGroup;
+
+    return (
+        <Modal
+            heading="Add Usergroup"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+                        name="usergroup"
+                        className={styles.usergroupList}
+                        options={usergroupResponse?.results}
+                        keySelector={usergroupKeySelector}
+                        labelSelector={usergroupLabelSelector}
+                        optionsPopupClassName={styles.optionsPopup}
+                        onChange={onValueChange}
+                        value={value.usergroup}
+                        error={error?.fields?.usergroup}
+                        placeholder="Select usergroup"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+                        name="role"
+                        className={styles.roleList}
+                        options={projectRolesResponse?.results}
+                        keySelector={roleKeySelector}
+                        labelSelector={roleLabelSelector}
+                        optionsPopupClassName={styles.optionsPopup}
+                        onChange={onValueChange}
+                        value={value.role}
+                        error={error?.fields?.role}
+                        placeholder="Select Role"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserGroupList/index.tsx [hunk 15]
```diff
@@ -38,21 +50,55 @@ function UserGroupList(props: Props) {
...
     const [
         userGroupPending,
         userGroupResponse,
+        ,
+        triggerUsergroupResponse,
     ] = useRequest<MultiResponse<UserGroup>>({
         url: `server://projects/${projectId}/project-usergroups/`,
         method: 'GET',
         query: queryForRequest,
         autoTrigger: true,
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/index.tsx:82]: handle failure

File: src/views/ProjectEdit/Users/UserGroupList/index.tsx [hunk 16]
```diff
@@ -38,21 +50,55 @@ function UserGroupList(props: Props) {
...
+        ,
+        ,
+        ,
+        triggerDeleteUsergroup,
+    ] = useRequest({
+        url: `server://projects/${projectId}/project-usergroups/${usergroupIdToDelete}/`,
+        method: 'DELETE',
+        onSuccess: () => {
+            triggerUsergroupResponse();
+        },
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserGroupList/index.tsx:98]: handle failure

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+        onValueChange,
+        validate,
+        onErrorSet,
+    } = useForm(userValue ?? defaultFormValue, schema);
+
+    const [
+        ,
+        projectRolesResponse,
+        ,
+        ,
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]: this is not necessary

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+        onFailure: (_, errorBody) => {
+            notifyOnFailure(_ts('projectEdit', 'projectMembershipPostFailed'))({ error: errorBody });
+        },
+    });
+
+    const handleSubmit = triggerAddProjectMember;
+
+    return (
+        <Modal
+            heading="Add Usergroup"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+                    <SelectInput
+                        name="member"
+                        className={styles.list}
+                        options={usersListResponse?.results}
+                        keySelector={membersKeySelector}
+                        labelSelector={membersLabelSelector}
+                        optionsPopupClassName={styles.optionsPopup}
+                        onChange={onValueChange}
+                        value={value.member}
+                        placeholder="Select User"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+                    <SelectInput
+                        name="role"
+                        className={styles.roleList}
+                        options={projectRolesResponse?.results}
+                        keySelector={roleKeySelector}
+                        labelSelector={roleLabelSelector}
+                        optionsPopupClassName={styles.optionsPopup}
+                        onChange={onValueChange}
+                        value={value.role}
+                        placeholder="Select Role"
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]: Use ts

File: src/views/ProjectEdit/Users/UserList/index.tsx [hunk 21]
```diff
@@ -36,21 +46,54 @@ function UserList(props: Props) {
...
     const [
         usersPending,
         usersResponse,
+        ,
+        triggerGetUsers,
     ] = useRequest<MultiResponse<Membership>>({
         url: `server://projects/${projectId}/project-memberships/`,
         method: 'GET',
         query: queryForRequest,
         autoTrigger: true,
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/index.tsx:76]: handle failure

File: src/views/ProjectEdit/Users/UserList/index.tsx [hunk 22]
```diff
@@ -36,21 +46,54 @@ function UserList(props: Props) {
...
+        ,
+        ,
+        triggerMembershipDelete,
+    ] = useRequest({
+        url: `server://projects/${projectId}/project-memberships/${membershipIdToDelete}/`,
+        method: 'DELETE',
+        onSuccess: () => {
+            triggerGetUsers();
+        },
+        autoTrigger: false,
```
Comment (AdityaKhatri) [inline src/views/ProjectEdit/Users/UserList/index.tsx:93]: handle failure

File: src/redux/initial-state/dev-lang.json [hunk 23]
```diff
@@ -1332,7 +1332,23 @@
...
+        "-93154523": "Failed to fetch users list",
+        "-56604597": "Delete usergroup from project",
+        "-76480274": "Delete user from project",
+        "-75535574": "Edit User",
+        "-70353397": "No usergroups added",
+        "-98959986": "No users added",
+        "-30147930": "Failed to Delete user membership",
+        "-53750920": "Select Role",
+        "-18211847": "Select User",
+        "-14111698": "Select User group",
```
Comment (tnagorra) [inline src/redux/initial-state/dev-lang.json]:
```suggestion
        "-14111698": "Select Usergroup",
```

File: src/views/ProjectEdit/Users/UserList/index.tsx [hunk 24]
```diff
@@ -113,6 +189,12 @@ function UserList(props: Props) {
         />
     ), []);
 
+    const membershipToEdit = useMemo(() => (
+        usersResponse?.results?.find(d => d.id === membershipIdToEdit)
+    ), [usersResponse?.results, membershipIdToEdit]);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/index.tsx:194]: Does this even work?

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+    });
+
+    const handleSubmit = useCallback(
+        () => {
+            const { errored, error: err, value: val } = validate();
+            onErrorSet(err);
+            if (!errored && isDefined(val)) {
+                setValueToSend(val as ValueToSend);
+            }
+            triggerAddProjectMember();
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]:
```suggestion
            if (!errored && isDefined(val)) {
                setValueToSend(val as ValueToSend);
                triggerAddProjectMember();
            }
```

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+    const formValueFromProps: PartialForm<FormType> = userValue;
+
+    const {
+        pristine,
+        value,
+        error,
+        onValueChange,
+        validate,
+        onErrorSet,
+    } = useForm(formValueFromProps ?? defaultFormValue, schema);
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx]:
```suggestion
    const formValueFromProps: PartialForm<FormType> = userValue ?? defaultFormValue;

    const {
        pristine,
        value,
        error,
        onValueChange,
        validate,
        onErrorSet,
    } = useForm(formValueFromProps, schema);
```

File: src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+};
+
+type FormSchema = ObjectSchema<PartialForm<FormType>>;
+type FormSchemaFields = ReturnType<FormSchema['fields']>
+
+const schema: FormSchema = {
+    fields: (value): FormSchemaFields => {
+        if (isDefined(value.id)) {
+            return ({
+                role: [requiredCondition],
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx:61]:
```suggestion
            return ({
                member: [clearCondition],
                role: [requiredCondition],
```

Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserList/AddUserModal/index.tsx:61]: Need to check if clearCondition should be added to member here!

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    const handleSubmit = useCallback(
+        () => {
+            const { errored, error: err, value: val } = validate();
+            onErrorSet(err);
+            if (!errored && isDefined(val)) {
+                setValueToSend(val as ValueToSend);
+            }
+            triggerAddUserGroup();
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx]:
```suggestion
            if (!errored && isDefined(val)) {
                setValueToSend(val as ValueToSend);
                triggerAddUserGroup();
            }
```

File: src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+};
+
+type FormSchema = ObjectSchema<PartialForm<FormType>>;
+type FormSchemaFields = ReturnType<FormSchema['fields']>;
+
+const schema: FormSchema = {
+    fields: (value): FormSchemaFields => {
+        if (isDefined(value.id)) {
+            return ({
+                role: [requiredCondition],
```
Comment (tnagorra) [inline src/views/ProjectEdit/Users/UserGroupList/AddUserGroupModal/index.tsx:67]: Need to confirm if usergroup clearCondition is required here

**Review by tnagorra — APPROVED** [review]
Need to fix submit logic

**Review by tnagorra — APPROVED** [review]

**Pull request merged**